core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1055,12 +1056,36 @@ private TrivialRule() {
       super(any(MutableRel.class), any(MutableRel.class), 0);
     }
 
-    @Override public @Nullable UnifyResult apply(UnifyRuleCall call) {
+    @Override
+    public @Nullable UnifyResult apply(UnifyRuleCall call) {
       if (call.query.equals(call.target)) {
         return call.result(call.target);
       }
+      if (isSpa(call.target) && MutableRels.descendants(call.query)
+          .containsAll(MutableRels.descendants(call.target))) {
+        return call.result(call.query);
+      }
       return null;
     }
+
+    private boolean isSpa(MutableRel rel) {
+      if (rel.getParent() != null) {
+        return false;
+      }
+      if (rel instanceof MutableCalc) {
+        MutableCalc topCalc = (MutableCalc) rel;
+        if (topCalc.getInput() instanceof MutableAggregate) {
+          MutableAggregate aggregate = (MutableAggregate) topCalc.getInput();
+          if (aggregate.getInput() instanceof MutableCalc) {
+            MutableCalc bottomCalc = (MutableCalc) aggregate.getInput();
+            if (bottomCalc.getInput() instanceof MutableScan) {
+              return true;
+            }
+          }
+        }
+      }
```

Comment:
Instead of this deep nested if, I think it might be easier to follow if you just returned false when any of these conditions fail and then everything could be flattened.